### PR TITLE
Decouples state handling from codec: completes baggage header example

### PR DESCRIPTION
**Pull request by adriancole**
This finishes integration of the baggage header example. We still have a
lot to think about before making dynamic baggage public. For example, we
had use cases such as tunneling over JMS (to use a single-header when
span.kind is messaging similar to what we do in B3Propagation).

At any rate, this is dramatically simplified in my opinion, and gets us
very close to completing both single header dynamic baggage and
secondary sampling.

See #693

### CODE REVIEW

File: brave/src/test/java/brave/features/baggage/BaggageInSingleHeaderTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.junit.Test;
+
+import static brave.baggage.BaggagePropagation.newFactoryBuilder;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.entry;
+
+/** This is an internal feature until we settle on an encoding format. */
+public class BaggageInSingleHeaderTest {
```
Comment (adriancole) [inline brave/src/test/java/brave/features/baggage/BaggageInSingleHeaderTest.java:32]: neat file 1:

File: brave/src/test/java/brave/features/baggage/SingleHeaderCodec.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This is a non-complete codec for the w3c (soon to be renamed to "baggage") header.
+ *
+ * <p>See https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md
+ */
+final class SingleHeaderCodec implements BaggageCodec {
```
Comment (adriancole) [inline brave/src/test/java/brave/features/baggage/SingleHeaderCodec.java:32]: neat file 2:

**Comment by adriancole** [comment]
I'm coming to a point of clarity on what to do with the baggage stuff vs secondary sampling:

1.   the extra state needed by secondary sampling shouldn't be forced through baggage apis. It was cleaner before
2. for baggage codec it is better to have only 2 options. something that handles 1 field and something that handles all fields
3. "local" fields must be marked as such as otherwise they are subverted by the all field codec (explicitly vs looking at key name count)
4. the ExtraBaggageFields type shouldn't be exposed to the codec. only the associated KV pairs
5. baggagepropagation internally should hide the <K> type as it is deprecated and always a string. we can special case when it isn't

File: brave/src/main/java/brave/internal/baggage/BaggageCodec.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+   * {@linkplain ExtraBaggageFields#isDynamic() dynamic values} will need to perform some decoding,
+   * such as splitting on comma and equals.
+   *
+   * @param <R> the type parameter of {@link Extractor#<R>}
+   * @param extra holds {@link BaggageField} state.
+   * @param request the parameter of {@link Extractor#extract(Object)}
+   * @param value a non-{@code null} result of {@link Getter#get(Object, Object)}
+   * @see #extractKeyNames()
+   */
+  <R> boolean decode(ExtraBaggageFields extra, R request, String value);
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/baggage/BaggageCodec.java:85]: agreed. this <R> drifted and isn't currently useful. I had one version that was passing the setter..

File: brave/src/main/java/brave/internal/baggage/DynamicBaggageState.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  @Override public boolean equals(Object o) {
+    if (o == this) return true;
+    if (!(o instanceof DynamicBaggageState)) return false;
+    return state.equals(((DynamicBaggageState) o).state);
+  }
+
+  @Override public int hashCode() {
+    return state.hashCode();
+  }
+}
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/baggage/DynamicBaggageState.java:78]: np. My formatter is actually correct now (I caused a lot of format problems when updating to an unknown version of intellij). regardless, I'll sweep for fuzz.. was a bit tired and didn't catch all of it last night.

File: brave/src/main/java/brave/internal/baggage/DynamicBaggageState.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      // a field value at some point (ex for reverting state).
+      LinkedHashMap<BaggageField, String> mergedState = new LinkedHashMap<>(state);
+      mergedState.put(field, value);
+      this.state = mergedState;
+    }
+    return true;
+  }
+
+  @Override void putAllIfAbsent(ExtraBaggageFields parent) {
+    Map<BaggageField, String> parentState = ((DynamicBaggageState) parent.internal).state;
```
[the remark this answers is not in the document]
Comment (adriancole) [reply brave/src/main/java/brave/internal/baggage/DynamicBaggageState.java:59]: you found a bug thanks!

**Comment by adriancole** [comment]
I'll take the 5 things noted into different PRs so this one doesn't become more blinding

**Comment by anuraaga** [comment]
LGTM

**Pull request merged**